compiler/rustc_infer/src/infer/canonical/canonicalizer.rs

```diff
@@ -778,4 +843,23 @@ impl<'cx, 'tcx> Canonicalizer<'cx, 'tcx> {
             )
         }
     }
+
+    /// Given a type variable `effect_var` of the given kind, first check
+    /// if `effect_var` is bound to anything; if so, canonicalize
+    /// *that*. Otherwise, create a new canonical variable for
+    /// `const_var`.
+    fn canonicalize_effect_var(
+        &mut self,
+        info: CanonicalVarInfo<'tcx>,
+        effect_var: ty::Effect<'tcx>,
+    ) -> ty::Effect<'tcx> {
+        let infcx = self.infcx;
+        let bound_to = infcx.shallow_resolve(effect_var);
+        if bound_to != effect_var {
```

Comment:
can you make that an assert instead? shouldn't we shallow resolve in `fold_effect` already?